icu4c/source/i18n/unicode/formattedvalue.h

```diff
@@ -288,7 +290,6 @@ class U_I18N_API FormattedValue /* not : public UObject because this is an inter
      */
     virtual Appendable& appendTo(Appendable& appendable, UErrorCode& status) const = 0;
 
-#ifndef U_HIDE_DRAFT_API
```

Comment:
> This was an intentional part of Steven's promotion PR. toString, toTempString, and appendTo were all on FormattedNumber from before, and only nextPosition was added new. So, we left nextPosition hidden both in the base class (here) and in the child class. Also note that un-hiding this may require un-hiding ConstrainedFieldPosition.

We cannot have #ifndef macros around virtual functions. That would mean clients would get different vtable layouts depending on which conditionals they build with, and lead to binary incompatibility. This is specifically called out in the [instructions for guard conditionals](https://sites.google.com/site/icusite/processes/release/tasks/docs#TOC-Verify-that-draft-is-surrounded-by-ifndef-U_HIDE_DRAFT_API-etc) and in the [C coding guidelines](http://userguide.icu-project.org/dev/codingguidelines#TOC-C-C-Hiding-Un--stable-APIs).

I guess I will un-hide ConstrainedFieldPosition.

Reply:
I wonder if that recommendation is still necessary. We had a similar issue when we put `std::atomic` as a class member in a public header. Despite the fact that the layout of `std::atomic` is implementation-dependent, we decided it was OK because we only guarantee binary compatibility in C APIs (C++ users build and ship ICU in tandem with the rest of their application).

But, it's not really a problem if you change it to expose nextPosition. As a way to avoid exposing ConstrainedFieldPosition, you could forward declare it since it's only used by reference in this function.

Reply:
It IS a problem because the library builds without anything hidden, but clients might build with say U_HIDE_DRAFT_API defined. It is not a matter of different binary versions of the library, but different builds of code that uses the SAME library version seeing different vtable layouts.

It is not necessary to really go out of our way to use the conditionals to hide everything that is DRAFT, say. The important thing is the API status tag. Then we conditionalize as much as we can easily do without causing any problems.